note on potential compilation error due to "derivative"

I got the following compilation error when I use both "ark-ff" and "stwo" in a repository.

```
error[E0433]: failed to resolve: could not find `clone` in `core`
  --> /Users/cusgadmin/.cargo/git/checkouts/stwo-e22b418082f2ccbd/2c8b6e5/crates/prover/src/core/poly/circle/evaluation.rs:18:10
   |
18 | #[derive(Derivative)]
   |          ^^^^^^^^^^ could not find `clone` in `core`
   |
   = note: this error originates in the derive macro `Derivative` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider importing this module
   |
1  + use std::clone;
   |

error[E0433]: failed to resolve: could not find `fmt` in `core`
  --> /Users/cusgadmin/.cargo/git/checkouts/stwo-e22b418082f2ccbd/2c8b6e5/crates/prover/src/core/poly/circle/evaluation.rs:18:10
   |
18 | #[derive(Derivative)]
   |          ^^^^^^^^^^ could not find `fmt` in `core`
   |
   = note: this error originates in the derive macro `Derivative` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider importing this module
   |
1  + use std::fmt;
   |

For more information about this error, try `rustc --explain E0433`.
```

The fundamental reason is that "derivative" is indeed a crate that is highly under-maintained, and it has many bugs that are present in previous generations of Rust. 

What happens is as follows:
- `ark-ff` will use the `use-core` feature of derivative. However, this causes an issue because this feature is propagated to other crates, as well.
- `derivative` is a crate that doesn't use the edition number, meaning that it is pre-2021. As a result, when it wants to use the core::clone::Clone, it uses ::core::clone:Clone. This would have behaved correctly post-2021 because "::" would direct it to the actual core crate. However, pre-2021, the behavior was not like this. Therefore, when stwo has a module also called "core", it actually redirects to stwo's core rather than the actual core.
- since the stwo's core module doesn't have fmt or clone, the derivative crate has the error.

There are a few solutions. 
- rename the module from "core" to something else
- ignore the ark-ff issue---someone would encounter it---but it would be very case-by-case
- replace derivative with some popular alternatives, such as https://crates.io/crates/educe. This one is getting more popularity nowadays with an all-time downloads 8m compared with derivative's 51m.

This issue is left as a note so that if someone else faces a similar issue, they can resolve it here.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -9,7 +9,7 @@ edition = "2021"
 [workspace.dependencies]
 blake2 = "0.10.6"
 blake3 = "1.5.0"
-derivative = "2.2.0"
+educe = "0.5.0"
 hex = "0.4.3"
 itertools = "0.12.0"
 num-traits = "0.2.17"

--- a/crates/prover/Cargo.toml
+++ b/crates/prover/Cargo.toml
@@ -10,7 +10,7 @@ blake2.workspace = true
 blake3.workspace = true
 bytemuck = { workspace = true, features = ["derive"] }
 cfg-if = "1.0.0"
-derivative.workspace = true
+educe.workspace = true
 hex.workspace = true
 itertools.workspace = true
 num-traits.workspace = true

--- a/crates/prover/src/core/lookups/mle.rs
+++ b/crates/prover/src/core/lookups/mle.rs
@@ -1,6 +1,6 @@
 use std::ops::Deref;
 
-use derivative::Derivative;
+use educe::Educe;
 
 use crate::core::backend::{Col, Column, ColumnOps};
 use crate::core::fields::qm31::SecureField;
@@ -15,8 +15,8 @@ pub trait MleOps<F: Field>: ColumnOps<F> + Sized {
 
 /// Multilinear Extension stored as evaluations of a multilinear polynomial over the boolean
 /// hypercube in bit-reversed order.
-#[derive(Derivative)]
-#[derivative(Debug(bound = ""), Clone(bound = ""))]
+#[derive(Educe)]
+#[educe(Debug, Clone)]
 pub struct Mle<B: ColumnOps<F>, F: Field> {
     evals: Col<B, F>,
 }

--- a/crates/prover/src/core/poly/circle/evaluation.rs
+++ b/crates/prover/src/core/poly/circle/evaluation.rs
@@ -1,7 +1,7 @@
 use std::marker::PhantomData;
 use std::ops::{Deref, Index};
 
-use derivative::Derivative;
+use educe::Educe;
 
 use super::{CanonicCoset, CircleDomain, CirclePoly, PolyOps};
 use crate::core::backend::cpu::CpuCircleEvaluation;
@@ -15,8 +15,8 @@ use crate::core::utils::bit_reverse_index;
 
 /// An evaluation defined on a [CircleDomain].
 /// The values are ordered according to the [CircleDomain] ordering.
-#[derive(Derivative)]
-#[derivative(Clone(bound = ""), Debug(bound = ""))]
+#[derive(Educe)]
+#[educe(Clone, Debug)]
 pub struct CircleEvaluation<B: FieldOps<F>, F: ExtensionOf<BaseField>, EvalOrder = NaturalOrder> {
     pub domain: CircleDomain,
     pub values: Col<B, F>,

--- a/crates/prover/src/core/vcs/prover.rs
+++ b/crates/prover/src/core/vcs/prover.rs
@@ -1,6 +1,7 @@
 use std::cmp::Reverse;
 use std::collections::BTreeMap;
 
+use educe::Educe;
 use itertools::Itertools;
 
 use super::ops::{MerkleHasher, MerkleOps};
@@ -202,7 +203,8 @@ impl<B: MerkleOps<H>, H: MerkleHasher> MerkleProver<B, H> {
     }
 }
 
-#[derive(Debug)]
+#[derive(Debug, Educe)]
+#[educe(Clone)]
 pub struct MerkleDecommitment<H: MerkleHasher> {
     /// Hash values that the verifier needs but cannot deduce from previous computations, in the
     /// order they are needed.
@@ -218,15 +220,6 @@ impl<H: MerkleHasher> MerkleDecommitment<H> {
         Self {
             hash_witness: Vec::new(),
             column_witness: Vec::new(),
-        }
-    }
-}
-// TODO(andreW): Remove these in favor of the `derivative` crate.
-impl<H: MerkleHasher> Clone for MerkleDecommitment<H> {
-    fn clone(&self) -> Self {
-        Self {
-            hash_witness: self.hash_witness.clone(),
-            column_witness: self.column_witness.clone(),
         }
     }
 }

